source/common/upstream/original_dst_cluster.cc

```diff
@@ -40,6 +41,26 @@ OriginalDstCluster::LoadBalancer::LoadBalancer(PrioritySet& priority_set, Cluste
 
 HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerContext* context) {
   if (context) {
+    // Check if request has a override host header set, if directly return the host using it.
+    const Http::HeaderMap* downstream_headers = context->downStreamHeaders();
+    if (downstream_headers &&
+        downstream_headers->get(Http::Headers::get().OriginalDstHostOverride) != nullptr) {
+      const std::string& request_override_host =
+          downstream_headers->get(Http::Headers::get().OriginalDstHostOverride)->value().c_str();
+      try {
+        Network::Address::InstanceConstSharedPtr overridehost_ip_port(
+            Network::Utility::parseInternetAddressAndPort(request_override_host, false));
+        ENVOY_LOG(debug, "Using request override host {}.", request_override_host);
+        return HostSharedPtr{new HostImpl(
+            info_, info_->name() + overridehost_ip_port->asString(),
+            std::move(overridehost_ip_port), envoy::api::v2::core::Metadata::default_instance(), 1,
+            envoy::api::v2::core::Locality().default_instance(),
+            envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())};
+      } catch (const Envoy::EnvoyException& e) {
+        ENVOY_LOG(warn, "original_dst_load_balancer: invalid override header value. {}", e.what());
```

Comment:
We shouldn't warn here, as this is in the request path and a bad client can spam the logs. I would make this debug level, and also add a stat for this (and tests for the stat, bad values, etc.).

Reply:
I have moved it to debug level. I did not add the stat for now as the existing log line below will indicate that. I think we can add in future if we see the need.